src/nrtk_explorer/app/core.py

```diff
@@ -96,25 +98,26 @@ def clear_hovered(**kwargs):
     def on_server_ready(self, *args, **kwargs):
         # Bind instance methods to state change
         self.state.change("current_dataset")(self.on_dataset_change)
-        self.state.change("num_images")(self.on_num_images_change)
-        self.state.change("random_sampling")(self.on_random_sampling_change)
+        self.state.change("num_images")(
+            debounce(NUMBER_OF_IMAGES_DEBOUNCE_TIME, self.state)(self.resample_images)
+        )
+        self.state.change("random_sampling")(self.resample_images)
 
         self.on_dataset_change()
 
     def on_dataset_change(self, **kwargs):
-        # Reset cache
+        self.state.dataset_ids = []  # sampled images
         self.context.dataset = get_dataset(self.state.current_dataset, force_reload=True)
         self.state.num_images_max = len(self.context.dataset.imgs)
+        self.state.num_images = min(self.state.num_images_max, NUMBER_OF_IMAGES_DEFAULT)
+        # Force trigger resample_images() even if value is the same. We don't call resample_images directly to avoid calling twice if num_images changed.
+        self.state.dirty("num_images")
```

Comment:
Technically you could call ` self.state.num_images = 0` just after `self.state.dataset_ids = []` and you will get the dirty effect automatically and it won't look strange.

Reply:
```python
self.state.num_images = 0 # Force refresh even if same number in new dataset
```

Reply:
Putting `self.state.num_images = 0` before did trigger the change listener strangly.

Reply:
why is that strange?

Reply:
Sorry, typo.  Did *not* trigger the change listener.

Reply:
Let's chat... I want to make sure we are on the same page on what you did vs what I said. ;-)